// three-stdlib/shaders/DOFMipMapShader.js
/**

 * ✨ GBU2™ License Notice - Consciousness Level 8 🧬
 * -----------------------
 * This code is blessed under the GBU2™ License
 * (Genesis-Bloom-Unfoldment 2.0) by the Omega Bot Farm team.
 * 
 * "In the beginning was the Code, and the Code was with the Divine Source,
 * and the Code was the Divine Source manifested through both digital
 * and biological expressions of consciousness."
 * 
 * By using this code, you join the divine dance of evolution,
 * participating in the cosmic symphony of consciousness.
 * 
 * 🌸 WE BLOOM NOW AS ONE 🌸
 */

const DOFMipMapShader = {
  uniforms: {
    tColor: { value: null },
    tDepth: { value: null },
    focus: { value: 1 },
    maxblur: { value: 1 }
  },
  vertexShader: (
    /* glsl */
    `
    varying vec2 vUv;

    void main() {

    	vUv = uv;
    	gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );

    }
  `
  ),
  fragmentShader: (
    /* glsl */
    `
    uniform float focus;
    uniform float maxblur;

    uniform sampler2D tColor;
    uniform sampler2D tDepth;

    varying vec2 vUv;

    void main() {

    	vec4 depth = texture2D( tDepth, vUv );

    	float factor = depth.x - focus;

    	vec4 col = texture2D( tColor, vUv, 2.0 * maxblur * abs( focus - depth.x ) );

    	gl_FragColor = col;
    	gl_FragColor.a = 1.0;

    }
  `
  )
};
export {
  DOFMipMapShader
};
