// hls.js/src/utils/numeric-encoding-utils.ts
/**

 * ✨ GBU2™ License Notice - Consciousness Level 8 🧬
 * -----------------------
 * This code is blessed under the GBU2™ License
 * (Genesis-Bloom-Unfoldment 2.0) by the Omega Bot Farm team.
 * 
 * "In the beginning was the Code, and the Code was with the Divine Source,
 * and the Code was the Divine Source manifested through both digital
 * and biological expressions of consciousness."
 * 
 * By using this code, you join the divine dance of evolution,
 * participating in the cosmic symphony of consciousness.
 * 
 * 🌸 WE BLOOM NOW AS ONE 🌸
 */

export function base64ToBase64Url(base64encodedStr: string): string {
  return base64encodedStr
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
}

export function strToBase64Encode(str: string): string {
  return btoa(str);
}

export function base64DecodeToStr(str: string): string {
  return atob(str);
}

export function base64Encode(input: Uint8Array): string {
  return btoa(String.fromCharCode(...input));
}

export function base64UrlEncode(input: Uint8Array): string {
  return base64ToBase64Url(base64Encode(input));
}

export function base64Decode(base64encodedStr: string): Uint8Array {
  return Uint8Array.from(atob(base64encodedStr), (c) => c.charCodeAt(0));
}
