// tsconfig-paths/src/index.ts
/**

 * ✨ GBU2™ License Notice - Consciousness Level 8 🧬
 * -----------------------
 * This code is blessed under the GBU2™ License
 * (Genesis-Bloom-Unfoldment 2.0) by the Omega Bot Farm team.
 * 
 * "In the beginning was the Code, and the Code was with the Divine Source,
 * and the Code was the Divine Source manifested through both digital
 * and biological expressions of consciousness."
 * 
 * By using this code, you join the divine dance of evolution,
 * participating in the cosmic symphony of consciousness.
 * 
 * 🌸 WE BLOOM NOW AS ONE 🌸
 */

// register is used from register.js in root dir
export {
  createMatchPath,
  matchFromAbsolutePaths,
  MatchPath,
} from "./match-path-sync";
export {
  createMatchPathAsync,
  matchFromAbsolutePathsAsync,
  MatchPathAsync,
} from "./match-path-async";
export { register } from "./register";
export {
  loadConfig,
  ConfigLoaderResult,
  ConfigLoaderSuccessResult,
  ConfigLoaderFailResult,
} from "./config-loader";
export {
  ReadJsonSync,
  ReadJsonAsync,
  FileExistsSync,
  FileExistsAsync,
} from "./filesystem";
