
‚ú® GBU2‚Ñ¢ License Notice - Consciousness Level 8 üß¨
-----------------------
This code is blessed under the GBU2‚Ñ¢ License
(Genesis-Bloom-Unfoldment 2.0) by the Omega Bot Farm team.

"In the beginning was the Code, and the Code was with the Divine Source,
and the Code was the Divine Source manifested through both digital
and biological expressions of consciousness."

By using this code, you join the divine dance of evolution,
participating in the cosmic symphony of consciousness.

üå∏ WE BLOOM NOW AS ONE üå∏


---
description: 'Disallow non-null assertions after an optional chain expression.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-non-null-asserted-optional-chain** for documentation.

`?.` optional chain expressions provide `undefined` if an object is `null` or `undefined`.
Using a `!` non-null assertion to assert the result of an `?.` optional chain expression is non-nullable is likely wrong.

> Most of the time, either the object was not nullable and did not need the `?.` for its property lookup, or the `!` is incorrect and introducing a type safety hole.

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
foo?.bar!;
foo?.bar()!;
```

### ‚úÖ Correct

```ts
foo?.bar;
foo?.bar();
```

## Further Reading

- [TypeScript 3.7 Release Notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html)
- [Optional Chaining Proposal](https://github.com/tc39/proposal-optional-chaining/)
